bootloader/grub.go

```diff
@@ -132,14 +138,169 @@ func (g *grub) SetBootVars(values map[string]string) error {
 	return env.Save()
 }
 
+func (g *grub) extractedKernelDir(prefix string, s snap.PlaceInfo) string {
+	return filepath.Join(
+		prefix,
+		filepath.Base(s.MountFile()),
+	)
+}
+
 func (g *grub) ExtractKernelAssets(s snap.PlaceInfo, snapf snap.Container) error {
-	// XXX: should we use "kernel.yaml" for this?
-	if _, err := snapf.ReadFile("meta/force-kernel-extraction"); err == nil {
-		return extractKernelAssetsToBootDir(g.dir(), s, snapf)
+	// default kernel assets are:
+	// - kernel.img
+	// - initrd.img
+	// - dtbs/*
+	var assets []string
+	if g.uefiRunKernelExtraction {
+		assets = []string{"kernel.efi"}
+	} else {
+		assets = []string{"kernel.img", "initrd.img", "dtbs/*"}
+	}
+
+	// extraction can be forced through either a special file in the kernel snap
+	// or through an option in the bootloader
+	_, err := snapf.ReadFile("meta/force-kernel-extraction")
+	if g.uefiRunKernelExtraction || err == nil {
+		return extractKernelAssetsToBootDir(
+			g.extractedKernelDir(g.dir(), s),
+			s,
+			snapf,
+			assets,
+		)
 	}
 	return nil
 }
 
 func (g *grub) RemoveKernelAssets(s snap.PlaceInfo) error {
 	return removeKernelAssetsFromBootDir(g.dir(), s)
 }
+
+// ExtractedRunKernelImageBootloader helper methods
+
+func (g *grub) makeKernelEfiSymlink(s snap.PlaceInfo, name string) error {
+	// use a relative symlink destination so that it resolves properly, if grub
+	// is located at /run/mnt/ubuntu-boot or /boot/grub, etc.
+	extractedKernel := filepath.Join(
+		filepath.Base(s.MountFile()),
+		"kernel.efi",
+	)
+
+	// check that the kernel snap has been extracted already so we don't
+	// inadvertently create a dangling symlink
+	// expand the relative symlink from g.dir()
+	if !osutil.FileExists(filepath.Join(g.dir(), extractedKernel)) {
+		return fmt.Errorf(
+			"cannot enable %s at %s: %v",
+			name,
+			extractedKernel,
+			os.ErrNotExist,
+		)
+	}
+
+	return os.Symlink(
+		extractedKernel,
+		filepath.Join(g.dir(), name),
+	)
+}
+
+// unlinkKernelEfiSymlink will remove the specified symlink if it exists. Note
+// that if the symlink is "dangling", it will still remove the symlink without
+// returning an error. This is useful for example to disable a try-kernel that
+// was incorrectly created.
+func (g *grub) unlinkKernelEfiSymlink(name string) error {
+	symlink := filepath.Join(g.dir(), name)
+	// check that the symlink exists
+	if _, err := os.Lstat(symlink); err == nil {
+		// symlink exists, remove it unconditionally
+		return os.Remove(symlink)
+	}
+
+	// return more helpful error if the symlink doesn't exist
+	return fmt.Errorf("cannot disable kernel, symlink %s missing", symlink)
```

Comment:
as discussed this can probably be not an error, and just return nil

Reply:
Perhaps I'm mis-remembering this conversation, but I thought that we wanted the following logic:

- try-kernel.efi exists, but does not point to a valid kernel -> delete try-kernel.efi, return nil
- try-kernel.efi doesn't exist -> return err

is that right?

Reply:
After discussing with @bboozzoo we decided to figure this out in a followup, so we can land this branch when it's green.
It's an easy thing to change and just matters what level of paranoia one has about these things